File: ObjectCDemo/ObjectCDemo/ViewController.m
//
//  ViewController.m
//  ObjectCDemo
//
//  Created by apple on 2023/1/9.
//

#import "ViewController.h"
#import <FuhaiRbox/FuhaiRbox-Swift.h>
#import "Masonry.h"

@interface ViewController ()
@property (nonatomic, strong) NSString *boxShowLog;
@property (nonatomic, strong) NSString *fh_macid;
@property (nonatomic, strong) NSString *fh_recycleId;
@property (nonatomic, strong) NSString *fh_md5;
@property (nonatomic, assign) NSInteger fh_time;

@property (nonatomic, strong) UILabel *debugLab;
@property (nonatomic, strong) UISwitch *debugSwitch;
@property (nonatomic, strong) UIButton *clearBtn;
@property (nonatomic, strong) UITextView *codeView;
@property (nonatomic, strong) UIButton *scanBtn;
@property (nonatomic, strong) UIButton *md5Btn;
@property (nonatomic, strong) UIButton *openBtn;
@end

@implementation ViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    // Do any additional setup after loading the view.
    [self addAllSubviews];
    [self addAllConstraints];
    [self addLogEvent];
}

- (void)viewWillAppear:(BOOL)animated {
    [super viewWillAppear:animated];
    [self configFRbox:NO];
}

- (void)dealloc {
    NSLog(@"****%s",__func__);
}


#pragma mark -  UI Component
// 获取安全区域
- (UIEdgeInsets)safeArea {
    UIEdgeInsets inset = UIEdgeInsetsMake(20, 0, 0, 0);
    if (@available(iOS 11.0, *)) {
        inset = self.view.safeAreaInsets;
        if (inset.bottom <= 0) {
            inset = UIEdgeInsetsMake(20, 0, 0, 0);
        }
    }
    return inset;
}

- (void)addAllSubviews {
    if (@available(iOS 13.0, *)) {
        self.view.backgroundColor = UIColor.systemBackgroundColor;
    }
    [self.view addSubview:self.debugSwitch];
    [self.view addSubview:self.debugLab];
    [self.view addSubview:self.codeView];
    [self.view addSubview:self.scanBtn];
    [self.view addSubview:self.md5Btn];
    [self.view addSubview:self.openBtn];
    [self.view addSubview:self.clearBtn];
}

- (void)addAllConstraints {
    [_codeView mas_makeConstraints:^(MASConstraintMaker *make) {
        make.top.offset(self.safeArea.top + 64);
        make.left.offset(10);
        make.right.offset(-10);
        make.height.offset(350);
    }];
    
    [_debugSwitch mas_makeConstraints:^(MASConstraintMaker *make) {
        make.top.equalTo(self.codeView.mas_bottom).offset(20);
        make.centerX.equalTo(self.view);
        make.width.offset(51);
        make.height.offset(31);
    }];

    [_debugLab mas_makeConstraints:^(MASConstraintMaker *make) {
        make.centerX.equalTo(self.view);
        make.top.equalTo(self.debugSwitch.mas_bottom).offset(5);
    }];

    [_scanBtn mas_makeConstraints:^(MASConstraintMaker *make) {
        make.top.equalTo(self.debugLab.mas_bottom).offset(20);
        make.left.offset(60);
        make.width.offset(100);
        make.height.offset(50);
    }];
    
    [_md5Btn mas_makeConstraints:^(MASConstraintMaker *make) {
        make.top.equalTo(self.debugLab.mas_bottom).offset(20);
        make.right.offset(-60);
        make.width.offset(100);
        make.height.offset(50);
    }];
    
    [_openBtn mas_makeConstraints:^(MASConstraintMaker *make) {
        make.top.equalTo(self.scanBtn.mas_bottom).offset(20);
        make.left.offset(60);
        make.width.offset(100);
        make.height.offset(50);
    }];
    
    [_clearBtn mas_makeConstraints:^(MASConstraintMaker *make) {
        make.top.equalTo(self.scanBtn.mas_bottom).offset(20);
        make.right.offset(-60);
        make.width.offset(100);
        make.height.offset(50);
    }];
}

- (UISwitch *)debugSwitch {
    if (!_debugSwitch) {
        _debugSwitch = [[UISwitch alloc]init];
        [_debugSwitch addTarget:self action:@selector(debugSwitch:) forControlEvents:UIControlEventValueChanged];
    }
    return _debugSwitch;
}

- (UILabel *)debugLab {
    if (!_debugLab) {
        _debugLab = [[UILabel alloc]init];
        _debugLab.font = [UIFont boldSystemFontOfSize:14];
        _debugLab.text = @"发布环境";
    }
    return _debugLab;
}

- (UITextView *)codeView {
    if (!_codeView) {
        _codeView= [[UITextView alloc]init];
        _codeView.backgroundColor = [[UIColor alloc]initWithWhite:0.5 alpha:0.1];
        _codeView.layer.cornerRadius = 8;
        _codeView.editable = NO;
        if (@available(iOS 15.0, *)) {
            [self setContentScrollView:nil forEdge:NSDirectionalRectEdgeNone];
        }
        return _codeView;
    }
    return _codeView;
}

- (UIButton *)clearBtn {
    if (!_clearBtn) {
        _clearBtn = [UIButton buttonWithType:UIButtonTypeSystem];
        _clearBtn.titleLabel.font = [UIFont boldSystemFontOfSize:14];
        [_clearBtn setTitle:@"清除日志" forState:UIControlStateNormal];
        [_clearBtn addTarget:self action:@selector(clearboxShowLog) forControlEvents:UIControlEventTouchUpInside];
    }
    return _clearBtn;
}

- (UIButton *)scanBtn {
    if (!_scanBtn) {
        _scanBtn = [UIButton buttonWithType:UIButtonTypeSystem];
        _scanBtn.titleLabel.font = [UIFont boldSystemFontOfSize:14];
        [_scanBtn setTitle:@"二维码扫描" forState:UIControlStateNormal];
        [_scanBtn addTarget:self action:@selector(scanFRbox) forControlEvents:UIControlEventTouchUpInside];
    }
    return _scanBtn;
}

- (UIButton *)md5Btn {
    if (!_md5Btn) {
        _md5Btn = [UIButton buttonWithType:UIButtonTypeSystem];
        _md5Btn.titleLabel.font = [UIFont boldSystemFontOfSize:14];
        [_md5Btn setTitle:@"获取MD5密码" forState:UIControlStateNormal];
        [_md5Btn addTarget:self action:@selector(requestMd5) forControlEvents:UIControlEventTouchUpInside];
    }
    return _md5Btn;
}

- (UIButton *)openBtn {
    if (!_openBtn) {
        _openBtn = [UIButton buttonWithType:UIButtonTypeSystem];
        _openBtn.titleLabel.font = [UIFont boldSystemFontOfSize:14];
        [_openBtn setTitle:@"Rbox开箱" forState:UIControlStateNormal];
        [_openBtn addTarget:self action:@selector(openFRbox) forControlEvents:UIControlEventTouchUpInside];
    }
    return _openBtn;
}


#pragma mark -  开箱子方法
- (void)clearboxShowLog {
    self.codeView.text = @"";
    self.boxShowLog = @"";
}

- (void)configFRbox:(BOOL)isDebug {
    NSString *appKey = @"queinier0va9yae2aimohchie4awooPh";
    NSString *appSecret = @"queisheewahnahNgoveeXai7iShei4ooyo3cu0hoh3ahsoo9Jeeshoh7Aif4Xee1";
    [FRbox configWithAppKey:appKey appSecret:appSecret isDebug:isDebug block:^(NSInteger code) {
    }];
}

- (void)scanFRbox {
    __weak typeof(self) weakSelf = self;
    [FRbox scanRboxWithBlock:^(NSInteger code, NSString *macid) {
        if (code == 200) {
            weakSelf.fh_macid = macid;
        }
    }];
}

- (void)openFRbox {
    if (_fh_md5.length > 0 && _fh_time > 0 && _fh_recycleId.length > 0) {
        __weak typeof(self) weakSelf = self;
        [FRbox openRboxWithMd5:_fh_md5 time:_fh_time recycleId:_fh_recycleId block:^(NSInteger code) {
            weakSelf.fh_time = 0;
            weakSelf.fh_md5 = nil;
            weakSelf.fh_recycleId = nil;
        }];
    }else {
        self.boxShowLog = [NSString stringWithFormat:@"%@请先获取Md5和时间戳和循环Id\n",self.boxShowLog];
        self.codeView.text = self.boxShowLog;
    }
}


// MARK: - 测试专用方法
- (void)addLogEvent {
    __weak typeof(self) weakSelf = self;
    [FRbox logEventWithBlock:^(NSString * msg) {
        weakSelf.boxShowLog = [NSString stringWithFormat:@"%@%@\n",self.boxShowLog,msg];
        weakSelf.codeView.text = weakSelf.boxShowLog;
    }];
}

- (void)debugSwitch:(UISwitch *)sender {
    if(sender.isOn) {
        self.debugLab.text = @"Debug环境";
    }else {
        self.debugLab.text = @"Release环境";
    }
    /// 重新验证授权
    [self configFRbox:sender.isOn];
}

- (void)requestMd5 {
    if (self.fh_macid.length > 0) {
        __weak typeof(self) weakSelf = self;
        [FRbox md5PwdWithMacId:_fh_macid block:^(NSInteger code, NSString *md5, NSInteger time, NSString *recycleId) {
            if (code == 200) {
                weakSelf.fh_time = time;
                weakSelf.fh_md5 = md5;
                weakSelf.fh_recycleId = recycleId;
            }
        }];
    }else {
        self.boxShowLog = [NSString stringWithFormat:@"%@请先扫描运单号获取MacId\n",self.boxShowLog];
        self.codeView.text = self.boxShowLog;
    }
}

@end

File: SwiftDemo/SwiftDemo/ViewController.swift
//
//  ViewController.swift
//  DemoTest
//
//  Created by yiyi on 2022/12/4.
//

import UIKit
import SnapKit
import FuhaiRbox

class ViewController: UIViewController {
    var boxShowLog: String = ""
    var fh_macid: String?
    var fh_time: Int?
    var fh_md5: String?
    var fh_recycleId: String?
    
    lazy var debugLab: UILabel = {
        let label = UILabel.init()
        label.font = .boldSystemFont(ofSize: 14)
        label.text = "发布环境"
        return label
    }()
    
    lazy var debugSwitch: UISwitch = {
        let deSwitch = UISwitch.init()
        deSwitch.isOn = false
        deSwitch.addTarget(self, action: #selector(debugSwitch(sender:)), for: .valueChanged)
        return deSwitch
    }()
    
    lazy var clearBtn: UIButton = {
        let button = UIButton.init(type: .system)
        button.titleLabel?.font = .boldSystemFont(ofSize: 14)
        button.setTitle("清除日志", for: .normal)
        button.addTarget(self, action: #selector(clearboxShowLog), for: .touchUpInside)
        return button
    }()
    
    lazy var codeView: UITextView = {
        let view = UITextView()
        view.backgroundColor = UIColor(white: 0.5, alpha: 0.1)
        view.layer.cornerRadius = 8
        view.isEditable = false
        if #available(iOS 15.0, *) {
            self.setContentScrollView(nil)
        }
        return view
    }()
    
    lazy var scanBtn: UIButton = {
        let button = UIButton.init(type: .system)
        button.titleLabel?.font = .boldSystemFont(ofSize: 14)
        button.setTitle("二维码扫描", for: .normal)
        button.addTarget(self, action: #selector(scanFRbox), for: .touchUpInside)
        return button
    }()
    
    lazy var md5Btn: UIButton = {
        let button = UIButton.init(type: .system)
        button.titleLabel?.font = .boldSystemFont(ofSize: 14)
        button.setTitle("获取MD5密码", for: .normal)
        button.addTarget(self, action: #selector(requestMd5), for: .touchUpInside)
        return button
    }()
    
    lazy var openBtn: UIButton = {
        let button = UIButton.init(type: .system)
        button.titleLabel?.font = .boldSystemFont(ofSize: 14)
        button.setTitle("Rbox开箱", for: .normal)
        button.addTarget(self, action: #selector(openFRbox), for: .touchUpInside)
        return button
    }()
    
    override func viewDidLoad() {
        super.viewDidLoad()
        // Do any additional setup after loading the view.
        addAllSubviews()
        addAllConstraints()
        addLogEvent()
    }

    override func viewWillAppear(_ animated: Bool) {
        super.viewWillAppear(animated)
        configFRbox()
    }
    
    deinit {
        print("***\(Self.self) \(#function)")
    }
    
    
    // MARK: - UI Component
    func addAllSubviews() {
        if #available(iOS 13.0, *) {
            view.backgroundColor = .systemBackground
        }
        view.addSubview(self.debugSwitch)
        view.addSubview(self.debugLab)
        view.addSubview(self.codeView)
        view.addSubview(self.scanBtn)
        view.addSubview(self.md5Btn)
        view.addSubview(self.openBtn)
        view.addSubview(self.clearBtn)
    }
    
    func addAllConstraints() {
        
        codeView.snp.makeConstraints { make in
            make.top.equalTo(safeArea.top + 64)
            make.left.equalTo(10)
            make.right.equalTo(-10)
            make.height.equalTo(350)
        }
        
        debugSwitch.snp.makeConstraints { make in
            make.top.equalTo(codeView.snp.bottom).offset(20)
            make.centerX.equalTo(self.view)
            make.width.equalTo(51)
            make.height.equalTo(31)
        }
    
        debugLab.snp.makeConstraints { make in
            make.centerX.equalTo(self.view)
            make.top.equalTo(debugSwitch.snp.bottom).offset(5)
        }

        scanBtn.snp.makeConstraints { make in
            make.top.equalTo(debugLab.snp.bottom).offset(20)
            make.left.equalTo(60)
            make.width.equalTo(100)
            make.height.equalTo(50)
        }
        
        md5Btn.snp.makeConstraints { make in
            make.top.equalTo(debugLab.snp.bottom).offset(20)
            make.right.equalTo(-60)
            make.width.equalTo(100)
            make.height.equalTo(50)
        }
        
        openBtn.snp.makeConstraints { make in
            make.top.equalTo(scanBtn.snp.bottom).offset(20)
            make.left.equalTo(60)
            make.width.equalTo(100)
            make.height.equalTo(50)
        }
        
        clearBtn.snp.makeConstraints { make in
            make.top.equalTo(scanBtn.snp.bottom).offset(20)
            make.right.equalTo(-60)
            make.width.equalTo(100)
            make.height.equalTo(50)
        }
        
    }
    
    
    // MARK: - 开箱子方法
    @objc func clearboxShowLog() {
        codeView.text = ""
        boxShowLog = ""
    }
    
    @objc func configFRbox(isDebug: Bool = false) {
        let appKey = "queinier0va9yae2aimohchie4awooPh"
        let appSecret = "queisheewahnahNgoveeXai7iShei4ooyo3cu0hoh3ahsoo9Jeeshoh7Aif4Xee1"
        FRbox.config(appKey:appKey, appSecret: appSecret, isDebug: isDebug){_ in }
    }
    
    @objc func scanFRbox() {
        FRbox.scanRbox {[weak self] code, data in
            if code == 200 {
                self?.fh_macid = data
            }
        }
    }
    
    @objc func openFRbox() {
        guard let md5 = fh_md5, let time = fh_time, let recycleId = fh_recycleId else {
            boxShowLog = boxShowLog + "请先获取Md5和时间戳和循环Id\n"
            codeView.text = boxShowLog
            return
        }
        FRbox.openRbox(md5: md5, time: time, recycleId: recycleId) {[weak self] _ in
            self?.fh_time = nil
            self?.fh_md5 = nil
            self?.fh_recycleId = nil
        }
    }
    
    
    // MARK: - 测试专用方法
    func addLogEvent() {
        FRbox.logEvent {[weak self] data in
            self?.boxShowLog = (self?.boxShowLog ?? "") + data + "\n"
            self?.codeView.text = self?.boxShowLog
        }
    }
    
    @objc func debugSwitch(sender: UISwitch) {
        if(sender.isOn) {
            debugLab.text = "Debug环境"
        }else {
            debugLab.text = "Release环境"
        }
        /// 重新验证授权
        configFRbox(isDebug: sender.isOn)
    }
    
    @objc func requestMd5() {
        guard let macid = fh_macid else {
            boxShowLog = boxShowLog + "请先扫描运单号获取MacId\n"
            codeView.text = boxShowLog
            return
        }
        FRbox.md5Pwd(macId: macid) {[weak self] code, md5, time, recycleId  in
            self?.fh_macid = nil
            if code == 200 {
                self?.fh_md5 = md5
                self?.fh_time = time
                self?.fh_recycleId = recycleId
            }
        }
    }
}


extension ViewController {
    // 获取安全区域
    var safeArea: UIEdgeInsets {
        var inset =  UIEdgeInsets(top: 20, left: 0, bottom: 0, right: 0)
        guard #available(iOS 11.0, *) else { return inset }
        inset = view.safeAreaInsets
        if inset.bottom <= 0 {
            inset = UIEdgeInsets(top: 20, left: 0, bottom: 0, right: 0)
        }
        return inset
    }
}

